pkg/trace/buffered_file.go

```diff
@@ -2,55 +2,100 @@ package trace
 
 import (
 	"bufio"
+	"errors"
+	"io"
 	"os"
 	"sync"
+	"sync/atomic"
 )
 
+// bufferedFile is a file that is being written to and read from. It is thread
+// safe, however, when reading from the file, writes will be ignored.
 type bufferedFile struct {
-	mut *sync.RWMutex
+	// reading protects the file from being written to while it is being read
+	// from. This is needed beyond in addition to the mutex so that writes can
+	// be ignored while reading.
+	reading atomic.Bool
+
+	// mut protects the buffered writer.
+	mut *sync.Mutex
+
 	// file is the file that is being written to.
 	file *os.File
+
 	// writer is the buffered writer that is writing to the file.
 	wr *bufio.Writer
 }
 
 // newbufferedFile creates a new buffered file that writes to the given file.
 func newbufferedFile(file *os.File) *bufferedFile {
 	return &bufferedFile{
-		mut:  &sync.RWMutex{},
-		file: file,
-		wr:   bufio.NewWriter(file),
+		file:    file,
+		wr:      bufio.NewWriter(file),
+		reading: atomic.Bool{},
+		mut:     &sync.Mutex{},
 	}
 }
 
-// Write writes the given bytes to the file.
+// Write writes the given bytes to the file. If the file is currently being read
+// from, the write will be lost.
 func (f *bufferedFile) Write(b []byte) (int, error) {
+	if f.reading.Load() {
+		return 0, nil
+	}
 	f.mut.Lock()
 	defer f.mut.Unlock()
 	return f.wr.Write(b)
 }
 
-// Flush flushes the writer to the file.
-func (f *bufferedFile) Flush() error {
+func (f *bufferedFile) startReading() error {
+	f.reading.Store(true)
+	f.mut.Lock()
+	defer f.mut.Unlock()
```

Comment:
good point, I'll take a look and see if we can simply this. The initial add for two different mutexes was so that we can ignore writes on line 43 w/o actually holding the mutex. Perhaps we can only use the reading atomic

Reply:
I tried a few things and unfortunately I think we'll have to keep both.

we don't want to only use the mutex, as that will block writes when reading, which could halt the node. If we only check if the mutex is locked then we risk ignoring writes even though we aren't reading. If we only use the reading atomic bool then we hit data races.

Reply:
> will block writes when reading

I thought that is intended to be this way based on the godoc for `bufferedFile`
```go
// bufferedFile is a file that is being written to and read from. It is thread
// safe, however, **when reading from the file, writes will be ignored**.
```
Maybe that comment needs some revisions?

Reply:
> >will block writes when reading

> I thought that is intended to be this way based on the godoc for bufferedFile

yeah we can definitely rephrase that. I think I'm some author bias unfortunately and need some help making it clearer, naively this part seems to indicate that writes are ignored. 
> **when reading from the file, writes will be ignored**.

perhaps a modification to the part on thread safety?